### DISCUSSION TITLE
Tool mode is dangerous, there should be separate script for editor only.

### DISCUSSION THREAD

**Discussion by SilencedPerson** [godotengine/godot-proposals#4306] [Engine Core]
Instead of having one script for both editor and the engine that shares code and can introduce bugs into your game, there should be separate script for editor logic and separate script for game logic. that way the logic stays separated, there is smaller chance of bugs occurring and it probably also has performance benefits.

On export/testing the scene, the editor code would be discarded in favor of the game script.

**Comment by Zireael07** [comment]
As someone who uses tool mode extensively, my vote is no. Yes, sometimes I need separate stuff for editor (e.g. debug drawing) but equally as often I'll need things to happen both in game and in editor, e.g. when generating game levels (doubly so as things spawned in editor by default don't have an owner set, meaning you don't see them in the scene tree and they don't serialize)

~~Also we don't have editor-only nodes so the point is moot because it's not possible to do what you want.~~ EDIT: misread

**Reply by SilencedPerson** [reply]
> As someone who uses tool mode extensively, my vote is no. Yes, sometimes I need separate stuff for editor (e.g. debug drawing) but equally as often I'll need things to happen both in game and in editor, e.g. when generating game levels (doubly so as things spawned in editor by default don't have an owner set, meaning you don't see them in the scene tree and they don't serialize)
> 
> Also we don't have editor-only nodes so the point is moot because it's not possible to do what you want.

The proposal is technically already implemented, you can achieve what i am proposing with set_script(). What you described as problems is incorrect.

only change required is to make this script replacement happen before the node being put into the tree.

**Reply by Zireael07** [reply]
Ah, I didn't notice that you wanted to replace just **scripts**, not nodes. Yep, that's possible, but I imagine even more prone to accidental bugs/breakage than tool scripts

**Reply by SilencedPerson** [reply]
Why would it be more prone to mistakes?

**Comment by fire-forge** [comment]
You can use `Engine.editor_hint` to check if the script is running in the editor or in a running game, so just put any editor-only or game-only code behind an `if Engine.editor_hint` check. That seems easier than having a separate script for editor and game, because if you want to have something happen in both cases you would have to write it twice.

**Reply by SilencedPerson** [reply]
Can you name some examples that need something to happen at the same time?
honestly can't really think of anything.

**Reply by Zireael07** [reply]
I gave you an example above: procedurally generating maps. I also often use tool scripts to enforce certain things, eg. mesh and collider sizes matching, or snapping thing to another thing... Those also need to happen both in editor and in game

**Reply by fire-forge** [reply]
@SilencedPerson Yes, lots of cases. For example, say you made a custom node for curved UI layouts (like a curved HBoxContainer) and you want it to run in the editor as well. If editor and in-game scripts had to be separate, you'd need 2 copies of the script, but with the current setup, you add `@tool` to the top and it just works.

**Reply by SilencedPerson** [reply]
If you are concerned about duplicit code, inheritance still exists.

**Comment by Jeremi360** [comment]
No! This is like Unity does it and it is one of many reasons I hate Unity
